File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/ffi/NativeRuntimeClientManager.kt
package io.nativeblocks.runtime.ffi

import io.nativeblocks.runtime.BuildConfig
import io.nativeblocks.runtime.api.NativeblocksEdition
import io.nativeblocks.runtime.api.util.SDKConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

internal class NativeRuntimeClientManager(
    private val instanceName: String,
    edition: NativeblocksEdition,
    http: HttpClient,
    cacheDir: String,
) {
    val environment = edition.toEngineEnvironment(instanceName)

    private val config = SdkConfig(
        version = BuildConfig.VERSION,
        platform = SDKConfig.SDK_PLATFORM,
    )

    private val runtime: NativeblocksRuntime =
        NativeblocksRuntime(environment, config, http, cacheDir)

    val frameClient: FrameClient = runtime.frameClient()
    val scaffoldClient: ScaffoldClient = runtime.scaffoldClient()
    val experimentClient: ExperimentClient = runtime.experimentClient()
    val localizationClient: LocalizationClient = runtime.localizationClient()
    val globalParameterClient: GlobalParameterClient = runtime.globalParameterClient()

    fun frameStateManager(): FrameStateManager = frameClient.stateManager()
    fun localizationStateManager(): LocalizationStateManager = localizationClient.stateManager()

    fun close() {
        runCatching { frameClient.close() }
        runCatching { scaffoldClient.close() }
        runCatching { experimentClient.close() }
        runCatching { localizationClient.close() }
        runCatching { globalParameterClient.close() }
        runCatching { runtime.close() }
        runCatching { disposeInstance(instanceName) }
    }
}

private fun NativeblocksEdition.toEngineEnvironment(instanceName: String): NativeblocksEnvironment {
    return when (this) {
        is NativeblocksEdition.Cloud -> NativeblocksEnvironment(
            instanceName = instanceName,
            endpoint = endpoint,
            apiKey = apiKey,
            developmentMode = developmentMode,
        )

        is NativeblocksEdition.Community -> NativeblocksEnvironment(
            instanceName = instanceName,
            endpoint = "",
            apiKey = "",
            developmentMode = false,
        )
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/ffi/NativeRuntimeClientManager.swift
import Foundation
import NativeblocksRuntimeFFI

internal final class NativeRuntimeClientManager {

    private let instanceName: String
    private let runtime: NativeblocksRuntime

    let environment: NativeblocksRuntimeFFI.NativeblocksEnvironment

    let frameClient: FrameClient
    let scaffoldClient: ScaffoldClient
    let experimentClient: ExperimentClient
    let localizationClient: LocalizationClient
    let globalParameterClient: GlobalParameterClient

    init(
        instanceName: String,
        edition: NativeblocksEdition,
        http: HttpClient,
        cacheDir: String
    ) throws {
        self.instanceName = instanceName
        self.environment = edition.toEngineEnvironment(instanceName: instanceName)

        let config = SdkConfig(
            version: SDKConfig.SDK_VERSION,
            platform: SDKConfig.SDK_PLATFORM
        )

        self.runtime = try NativeblocksRuntime(
            environment: environment,
            config: config,
            http: http,
            cacheDir: cacheDir
        )

        self.frameClient = runtime.frameClient()
        self.scaffoldClient = runtime.scaffoldClient()
        self.experimentClient = runtime.experimentClient()
        self.localizationClient = runtime.localizationClient()
        self.globalParameterClient = runtime.globalParameterClient()
    }

    func frameStateManager() -> FrameStateManager {
        return frameClient.stateManager()
    }

    func localizationStateManager() -> LocalizationStateManager {
        return localizationClient.stateManager()
    }

    func close() {
        disposeInstance(instanceName: instanceName)
    }
}

extension NativeblocksEdition {
    fileprivate func toEngineEnvironment(instanceName: String) -> NativeblocksRuntimeFFI.NativeblocksEnvironment {
        switch self {
        case .cloud(let endpoint, let apiKey, let developmentMode):
            return NativeblocksRuntimeFFI.NativeblocksEnvironment(
                instanceName: instanceName,
                endpoint: endpoint,
                apiKey: apiKey,
                developmentMode: developmentMode
            )
        case .community:
            return NativeblocksRuntimeFFI.NativeblocksEnvironment(
                instanceName: instanceName,
                endpoint: "",
                apiKey: "",
                developmentMode: false
            )
        }
    }
}
